openedx/core/djangoapps/oauth_dispatch/docs/decisions/0009-jwt-in-session-cookie.rst

```diff
@@ -0,0 +1,177 @@
+9. Transport JWT in Session Cookie
```

Comment:
# Notes from internal review on Aug 6, 2018

## Accessing User Info

* Since the JWT is not accessible to the microfrontend with JWT Cookies, how should the microfrontend get user information? 
* Some options are:
	1. Add an extra round trip to get the user-data from a backend API, and then cache it in HTML5 Storage.
	2. Continue to use the current JS-accessible user-info cookie, which contain user-data.
	3. Have the server populate the initial DOM with this data, but only works for server-generated HTML.
* Note: JS code can access user-data (with any of the options above), unless CSP is used. 
* Note: we will tackle this issue when we soon work on the shared header-footer.

## CORS section
* Need to make note that 'withCredentials' attribute must also be set.
* This means that we will need to send all edX cookies even when we make cross-edx-domain AJAX calls.

## Cookie Performance and Limits

* A quick check shows that today we send 39 cookies of 3MB total size on edX calls.
* Nice-to-have: a dashboard to monitor cookie usage (number and size) on edX.org.
* Performance: we already send all cookies in our requests anyway, so this change will add to it, but not make matters significantly worse.

Reply:
Minor edit note: 3K, not 3MB of cookie size right now for edX calls. 4K is the practical limit.

Reply:
Looking at decision record 0003 (using JWT in the first place):

> ### Pluses
> * The long-term design of the system will be simpler by using simpler protocols and frameworks, such as JWT as access tokens.
> * **OAuth Clients obtain basic identity information within the JWT access token without needing to hit an extra user info endpoint.**
> * **Any microservice can validate the JWT as an assertion without making an extra round trip to the LMS.**
> * Although there is no RFC or IETF standard for our use of OAuth+JWT, we are using a relatively > maintained and used open source library for our implementation.
> ### Minuses
> * **Token invalidation and single Logout become more difficult.**
> * During the transition period, there will be multiple implementations, which may result in confusion and a more complex system. The shorter we keep the transition period, the better.

We're now at a point where browsers at least cannot get basic identity from the JWT (because it's hidden from JS using cookie storage). Furthermore, if we decide to solve the token invalidation problem in a robust way, we'll also be giving up the "no extra round trip" advantage, since any validation against a blacklist is going to require a roundtrip to something.

Given that, I wonder if we're at a place where we should just be giving the client a simple, small, randomly generated session key, and do the translation to JWT at the API gateway or middleware layers somehow. Invalidation becomes simpler, we still get the benefits of JWT from the point of view of the services, and we lower the risk that we'll leak the key somehow because of inappropriate crypto settings.

Reply:
My fly-by 2 cents, as the team decides:

The recent advances in asymmetric JWTs allow us to move away from including all IDAs in our web of trust. So having a middleware in each IDA that converts a session key to a JWT isn’t ideal from that perspective.  Additionally, self-contained user-data within the JWT allows a frontend to introduce a new user to a backend, without any additional round trips.

I view the API gateway as an additional layer of protection and *optional* single-point of access.  We need a strong-enough story that works without it for our core need of SSO for Frontends.  Once we do use the gateway, it can enhance our security story even further (with the tradeoff of having a centralized access point).

Today, our web frontends use Session-Auth to access all backend APIs.  That already has a similar invalidation story to JWT Cookies, and has been sufficient to date. Given that, I consider the JWT + its Refresh token a robust-enough solution with sufficient knobs (expiration dates and possibility of future blacklist) for us to tweak over time. The blacklist (if we really really need it) can be configured at a gateway or via distributed config.

Reply:
Oy, stop working on your vacation! :P

> The recent advances in asymmetric JWTs allow us to move away from including all IDAs in our web of trust. So having a middleware in each IDA that converts a session key to a JWT isn’t ideal from that perspective.

Ah, sorry, I wasn't clear. I definitely agree that the middleware shouldn't be creating and signing JWTs itself. But if the middleware delegates to another service (perhaps our elusive auth/user IDA), I don't think it expands our web of trust any. There's still only one entity that has the private key, and that's the provider generating the JWT. Individual IDAs still verify using public keys. Though I guess it does set us up to be awkwardly split into two worlds where inter-service calls use one method and end users use another.

> Today, our web frontends use Session-Auth to access all backend APIs. That already has a similar invalidation story to JWT Cookies, and has been sufficient to date.

I guess I'm not as familiar with the invalidation story with Session-Auth, but I thought we had the ability to ban users via the user standing middleware check, in which case the effect is immediate?

> I view the API gateway as an additional layer of protection and optional single-point of access. We need a strong-enough story that works without it for our core need of SSO for Frontends. Once we do use the gateway, it can enhance our security story even further (with the tradeoff of having a centralized access point).

That makes total sense to me.

> Additionally, self-contained user-data within the JWT allows a frontend to introduce a new user to a backend, without any additional round trips.
> ...
> Given that, I consider the JWT + its Refresh token a robust-enough solution with sufficient knobs (expiration dates and possibility of future blacklist) for us to tweak over time. The blacklist (if we really really need it) can be configured at a gateway or via distributed config.

Maybe I just don't understand the use cases well enough, but if somebody is doing something bad enough where we want to revoke the token or block the user, don't we want to be able to respond to it almost immediately instead of waiting an hour for a token to expire? I guess my concern is that the moment we add any knobs (e.g. blacklist, user ban, etc.) that could give us that kind of immediate-response capability, we lose the "without any additional round trips" advantage. I suppose we could turn the expiry knob way down to a short number of minutes, but that seems like it'd be painful from an end-user pain point of view, since they'd have to constantly see 401s and hit the refresh endpoint.

Reply:
To be clear, I'm not saying that JWT sessions can't work. It can definitely work. My concern is more that by the time it's doing everything we want it to do, we'll be giving up on most of its selling points, while continuing to be stuck with its drawbacks (e.g. pushing our cookie-size envelope, high refresh rate, etc.)

I guess most of my concern comes up around token invalidation. If it's the conclusion of the security working group that immediate invalidation isn't necessary, then most of my concerns are moot.